File: too_many_cooks_data/test/authentication_test.dart
/// Tests for agent authentication.
library;

import 'dart:js_interop';

import 'package:dart_node_core/dart_node_core.dart';
import 'package:nadz/nadz.dart';
import 'package:test/test.dart';
import 'package:too_many_cooks_data/too_many_cooks_data.dart';

extension type _Fs(JSObject _) implements JSObject {
  external void unlinkSync(String path);
  external bool existsSync(String path);
}

final _Fs _fs = _Fs(requireModule('fs') as JSObject);

void _deleteIfExists(String path) {
  try {
    if (_fs.existsSync(path)) {
      _fs.unlinkSync(path);
    }
  } on Object catch (_) {}
}

void main() {
  const testDbPath = '.test_authentication.db';
  TooManyCooksDb? db;

  setUp(() {
    _deleteIfExists(testDbPath);
    final config = createDataConfig(dbPath: testDbPath);
    final result = createDb(config);
    expect(result, isA<Success<TooManyCooksDb, String>>());
    db = (result as Success<TooManyCooksDb, String>).value;
  });

  tearDown(() {
    db?.close();
    _deleteIfExists(testDbPath);
  });

  test('authenticate succeeds with valid credentials', () {
    final regResult = db!.register('auth-agent');
    final reg = (regResult as Success<AgentRegistration, DbError>).value;

    final authResult = db!.authenticate(reg.agentName, reg.agentKey);
    expect(authResult, isA<Success<AgentIdentity, DbError>>());
    final agent = (authResult as Success<AgentIdentity, DbError>).value;
    expect(agent.agentName, 'auth-agent');
  });

  test('authenticate fails with invalid key', () {
    db!.register('auth-agent2');

    final authResult = db!.authenticate('auth-agent2', 'wrong-key');
    expect(authResult, isA<Error<AgentIdentity, DbError>>());
    final error = (authResult as Error<AgentIdentity, DbError>).error;
    expect(error.code, errUnauthorized);
  });

  test('authenticate fails for nonexistent agent', () {
    final authResult = db!.authenticate('nonexistent', 'any-key');
    expect(authResult, isA<Error<AgentIdentity, DbError>>());
    final error = (authResult as Error<AgentIdentity, DbError>).error;
    expect(error.code, errUnauthorized);
  });

  test('authenticate updates last_active timestamp', () {
    final regResult = db!.register('timestamp-agent');
    final reg = (regResult as Success<AgentRegistration, DbError>).value;

    final firstAuth = db!.authenticate(reg.agentName, reg.agentKey);
    final firstAgent = (firstAuth as Success<AgentIdentity, DbError>).value;

    // Small delay to ensure timestamp changes
    final secondAuth = db!.authenticate(reg.agentName, reg.agentKey);
    final secondAgent = (secondAuth as Success<AgentIdentity, DbError>).value;

    expect(secondAgent.lastActive, greaterThanOrEqualTo(firstAgent.lastActive));
  });
}

File: too-many-cooks/test/authentication_test.ts
/// Tests for agent authentication.

import { describe, it, beforeEach, afterEach } from "node:test";
import assert from "node:assert";

import fs from "node:fs";

import {
  type TooManyCooksDb,
  createDataConfig,
  createDb,
  ERR_UNAUTHORIZED,
} from "../lib/src/data/data.js";

const TEST_DB_PATH = ".test_authentication.db" as const;

const deleteIfExists = (path: string): void => {
  try {
    if (fs.existsSync(path)) {
      fs.unlinkSync(path);
    }
  } catch {
    // ignore
  }
};

describe("authentication", () => {
  let db: TooManyCooksDb | undefined;

  beforeEach(() => {
    deleteIfExists(TEST_DB_PATH);
    const config = createDataConfig({ dbPath: TEST_DB_PATH });
    const result = createDb(config);
    assert.strictEqual(result.ok, true);
    if (result.ok) {
      db = result.value;
    }
  });

  afterEach(() => {
    db?.close();
    deleteIfExists(TEST_DB_PATH);
  });

  it("authenticate succeeds with valid credentials", () => {
    const regResult = db!.register("auth-agent");
    assert.strictEqual(regResult.ok, true);
    if (!regResult.ok) {return;}
    const reg = regResult.value;

    const authResult = db!.authenticate(reg.agentName, reg.agentKey);
    assert.strictEqual(authResult.ok, true);
    if (!authResult.ok) {return;}
    const agent = authResult.value;
    assert.strictEqual(agent.agentName, "auth-agent");
  });

  it("authenticate fails with invalid key", () => {
    db!.register("auth-agent2");

    const authResult = db!.authenticate("auth-agent2", "wrong-key");
    assert.strictEqual(authResult.ok, false);
    if (!authResult.ok) {
      assert.strictEqual(authResult.error.code, ERR_UNAUTHORIZED);
    }
  });

  it("authenticate fails for nonexistent agent", () => {
    const authResult = db!.authenticate("nonexistent", "any-key");
    assert.strictEqual(authResult.ok, false);
    if (!authResult.ok) {
      assert.strictEqual(authResult.error.code, ERR_UNAUTHORIZED);
    }
  });

  it("authenticate updates last_active timestamp", () => {
    const regResult = db!.register("timestamp-agent");
    assert.strictEqual(regResult.ok, true);
    if (!regResult.ok) {return;}
    const reg = regResult.value;

    const firstAuth = db!.authenticate(reg.agentName, reg.agentKey);
    assert.strictEqual(firstAuth.ok, true);
    if (!firstAuth.ok) {return;}
    const firstAgent = firstAuth.value;

    // Small delay to ensure timestamp changes
    const secondAuth = db!.authenticate(reg.agentName, reg.agentKey);
    assert.strictEqual(secondAuth.ok, true);
    if (!secondAuth.ok) {return;}
    const secondAgent = secondAuth.value;

    assert.ok(secondAgent.lastActive >= firstAgent.lastActive);
  });
});
